### ISSUE TITLE
Filesystem full due to  /bitnami/postgresql/data/pg_wal/ 

### ISSUE THREAD

**Issue by DanPM** [bitnami/bitnami-docker-postgresql-repmgr#52]
Hello, I'm using 2 VMs in master-slave configuration, s below. 
Data size in folder /bitnami/postgresql/data/pg_wal/ increases constantly. 
Archiving seems to work as currently I have 375 files in "/archive_status", and 377 files in "/pg_wal".
Disabling slave and running only master doesn't help, files keep pilling up in /pg_wal.

Any hint would be really helpful.

```
$ docker run --detach --name pg-0 \
  -p 0.0.0.0:5432:5432 \
  --env REPMGR_PARTNER_NODES=10.49.6.49,10.49.6.50 \
  --env REPMGR_NODE_NAME=pg-0 \
  --env REPMGR_NODE_NETWORK_NAME=127.0.0.1 \
  --env REPMGR_PRIMARY_HOST=127.0.0.1 \
  --env REPMGR_PORT_NUMBER=5432 \
  --env REPMGR_PRIMARY_PORT=5432 \
  --env REPMGR_LOG_LEVEL=DEBUG \
  --env REPMGR_PASSWORD=password \
  --env POSTGRESQL_PASSWORD=password \
  -v /usr/docker-configs/postgresql/pg_database:/bitnami/postgresql \
  bitnami/postgresql-repmgr:latest

$ docker run --detach --name pg-1 \
  -p 0.0.0.0:5432:5432 \
  --env REPMGR_PARTNER_NODES=10.49.6.49,10.49.6.50 \
  --env REPMGR_NODE_NAME=pg-1 \
  --env REPMGR_NODE_NETWORK_NAME=127.0.0.1 \
  --env REPMGR_PRIMARY_HOST=10.49.6.49 \
  --env REPMGR_PORT_NUMBER=5432 \
  --env REPMGR_PRIMARY_PORT=5432 \
  --env REPMGR_LOG_LEVEL=DEBUG \
  --env REPMGR_PASSWORD=password \
  --env POSTGRESQL_PASSWORD=password \
  -v /usr/docker-configs/postgresql/pg_database:/bitnami/postgresql \
  bitnami/postgresql-repmgr:latest
```

**Comment by marcosbc** [comment]
Hi @DanPM, PostgreSQL WAL is used for ensuring data integrity, with the downside that it logs every transaction (meaning it requires significant disk space).

In the case of repmgr, it requires it to be set to either logical/hot_standby (we set it to the latter). Read more [here](https://repmgr.org/docs/current/configuration-prerequisites.html). You could try to set it to logical to see if it helps with the disk space problem, but in your case I would rather try and increase the disk to avoid issues, as PostgreSQL+repmgr requires WAL to be active anyways.

**Comment by DanPM** [comment]
Hi @marcosbc,

thank you very much for reply, and thank you for kind explanations (i'm 0 regarding DB/ DBA topics).

> Edit: By the way, could you share the size of your pg_wal directory?
Sorry didn't take a screenshot, but was almost 30GB on master node, and a couple of MB on slave.

I'm currently in the middle of loadtesting, so I can't say for sure if I solved the issue, but I did the followings:

1. Did some debugging and noticed slave node was not attached, and couldn't be attached:
```
docker exec -ti pg_slave /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr node check -f /opt/bitnami/repmgr/conf/repmgr.conf
docker exec -ti pg_slave /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr node status -f /opt/bitnami/repmgr/conf/repmgr.conf
docker exec -ti pg_slave /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr standby register --force  -f /opt/bitnami/repmgr/conf/
```
Although failover was working, clearly I had some setup issues.

2. I read the following article about password in _conninfo_ 
https://github.com/2ndQuadrant/repmgr/issues/509#issuecomment-459629793
 
3. downgraded to image tag 9.6 and slave was successfully attached to master

Since I downgraded, I don't see  the /pg_xlogs folder growing (and is also replicated to slave, same size ~300MB). 
As I said, I'm currently load tested so can't be sure if my issue is solved.


**Couple of questions:**
1. If this is true: https://github.com/2ndQuadrant/repmgr/issues/509#issuecomment-459629793 -> why I'm the only one complaining? everybody else should also notice the issue already.

2. If slave is not attached to master, might we conclude that is a replication issue with WAL files (which lead to filling up the filesystem)?

**Comment by dani8art** [comment]
Hi @DanPM thank you so much for a great explanation, I was playing a bit with the image using docker and helm and I was able to reproduce the issue in both. It seems the standby node is not properly attached to the cluster as we can see in the logs.

```console
$ docker exec -ti postgresql-repmgr-g731581598_pg-0_1 /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr cluster show -f /opt/bitnami/repmgr/conf/repmgr.conf
postgresql-repmgr 09:47:02.14
postgresql-repmgr 09:47:02.14 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 09:47:02.14 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 09:47:02.15 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 09:47:02.15
postgresql-repmgr 09:47:02.15 DEBUG ==> Configuring libnss_wrapper...

WARNING: node "pg-1" not found in "pg_stat_replication"
 ID   | Name | Role    | Status    | Upstream | Location | Priority | Timeline | Connection string
------+------+---------+-----------+----------+----------+----------+----------+-----------------------------------------------------------------------------------------
 1000 | pg-0 | primary | * running |          | default  | 100      | 1        | user=repmgr password=repmgrpassword host=pg-0 dbname=repmgr port=5432 connect_timeout=5
 1001 | pg-1 | standby |   running | ! pg-0   | default  | 100      | 1        | user=repmgr password=repmgrpassword host=pg-1 dbname=repmgr port=5432 connect_timeout=5

WARNING: following issues were detected
  - node "pg-1" (ID: 1001) is not attached to its upstream node "pg-0" (ID: 1000)
```

I did some test around the `repmgr standby register` action in our logic and it seems there could be two problems there:

1. we are shutting down postgresql before the wallreceiver process is executed
2. if we wait for this to happen, a connection error is received 

So that the standby node is not properly registered for the first time and it goes to an unstable state. 

**if we don't wait a shut down is performed before something happen**

```console
postgresql-repmgr 09:41:27.62 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down....2020-11-04 09:41:27.632 GMT [177] LOG:  received smart shutdown request
2020-11-04 09:41:27.634 GMT [197] FATAL:  terminating walreceiver process due to administrator command
2020-11-04 09:41:27.636 GMT [180] LOG:  shutting down
2020-11-04 09:41:27.651 GMT [177] LOG:  database system is shut down
 done
server stopped
```

**if we wait just adding `sleep 60` to the `setup.sh` script it receives a connection error and then shut down**

```console
postgresql-repmgr 09:40:27.57 INFO  ==> Registering Standby node...
NOTICE: using provided configuration file "/opt/bitnami/repmgr/conf/repmgr.conf"
WARNING: --upstream-node-id not supplied, assuming upstream node is primary (node ID 1000)
WARNING: node "pg-1" not found in "pg_stat_replication"
WARNING: local node not attached to primary node 1000
NOTICE: -F/--force supplied, continuing anyway
[REPMGR EVENT] Node id: 1001; Event type: standby_register; Success [1|0]: 1; Time: 2020-11-04 09:40:27.601108+00;  Details: standby registration succeeded; upstream node ID is 1000 (-F/--force option was used)
Looking for the script: /opt/bitnami/repmgr/events/execs/standby_register.sh
[REPMGR EVENT] no script '/opt/bitnami/repmgr/events/execs/standby_register.sh' found. Skipping...
NOTICE: standby node "pg-1" (ID: 1001) successfully registered
2020-11-04 09:41:19.072 GMT [183] LOG:  replication terminated by primary server
2020-11-04 09:41:19.072 GMT [183] DETAIL:  End of WAL reached on timeline 1 at 0/5000098.
2020-11-04 09:41:19.072 GMT [183] FATAL:  could not send end-of-streaming message to primary: no COPY in progress
2020-11-04 09:41:19.072 GMT [179] LOG:  invalid record length at 0/5000098: wanted 24, got 0
2020-11-04 09:41:19.075 GMT [196] FATAL:  could not connect to the primary server: could not connect to server: Connection refused
		Is the server running on host "pg-0" (172.25.0.3) and accepting
		TCP/IP connections on port 5432?
2020-11-04 09:41:24.081 GMT [197] LOG:  started streaming WAL from primary at 0/5000000 on timeline 1
postgresql-repmgr 09:41:27.62 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down....2020-11-04 09:41:27.632 GMT [177] LOG:  received smart shutdown request
2020-11-04 09:41:27.634 GMT [197] FATAL:  terminating walreceiver process due to administrator command
2020-11-04 09:41:27.636 GMT [180] LOG:  shutting down
2020-11-04 09:41:27.651 GMT [177] LOG:  database system is shut down
 done
server stopped
```

**Comment by DanPM** [comment]
Hi @dani8art ,

cool that you managed to replicate the issues, still is not clear to me why the connection error.

**Comment by dani8art** [comment]
> cool that you managed to replicate the issues, still is not clear to me why the connection error.

I was investigating a bit more and I was able to fix the issue, it seems like there is a race condition while creating the containers so the standby node cannot properly be attached to the primary one. If you wait for the primary to be ready to accept connections them the cluster is set correctly. I did the following.

Using https://raw.githubusercontent.com/bitnami/bitnami-docker-postgresql-repmgr/master/docker-compose.yml, I mark all the thing related to `pg-1` commented and just deploy `pg-0`.

```console
$ docker-compose up -d
Creating postgresql-repmgr-g731581598_pg-0_1 ... done
$ docker ps
CONTAINER ID        IMAGE                                    COMMAND                  CREATED             STATUS              PORTS                     NAMES
85199739923f        bitnami/postgresql-repmgr:11-debian-10   "/opt/bitnami/script…"   7 seconds ago       Up 6 seconds        0.0.0.0:32806->5432/tcp   postgresql-repmgr-g731581598_pg-0_1
```

When `pg-0` was configured an it was ready to accept connections I added `pg-1`.

```console
$ docker-compose up -d
postgresql-repmgr-g731581598_pg-0_1 is up-to-date
Creating postgresql-repmgr-g731581598_pg-1_1 ... done
$ docker ps
CONTAINER ID        IMAGE                                    COMMAND                  CREATED             STATUS              PORTS                     NAMES
e17eb09af024        bitnami/postgresql-repmgr:11-debian-10   "/opt/bitnami/script…"   3 seconds ago       Up 2 seconds        0.0.0.0:32807->5432/tcp   postgresql-repmgr-g731581598_pg-1_1
85199739923f        bitnami/postgresql-repmgr:11-debian-10   "/opt/bitnami/script…"   41 seconds ago      Up 40 seconds       0.0.0.0:32806->5432/tcp   postgresql-repmgr-g731581598_pg-0_1
```

The cluster was set correctly

```console
$ docker exec -ti postgresql-repmgr-g731581598_pg-0_1 /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr cluster show -f /opt/bitnami/repmgr/conf/repmgr.conf
postgresql-repmgr 09:27:00.60
postgresql-repmgr 09:27:00.60 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 09:27:00.60 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 09:27:00.60 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 09:27:00.60
postgresql-repmgr 09:27:00.61 DEBUG ==> Configuring libnss_wrapper...

 ID   | Name | Role    | Status    | Upstream | Location | Priority | Timeline | Connection string
------+------+---------+-----------+----------+----------+----------+----------+-----------------------------------------------------------------------------------------
 1000 | pg-0 | primary | * running |          | default  | 100      | 1        | user=repmgr password=repmgrpassword host=pg-0 dbname=repmgr port=5432 connect_timeout=5
 1001 | pg-1 | standby |   running | pg-0     | default  | 100      | 1        | user=repmgr password=repmgrpassword host=pg-1 dbname=repmgr port=5432 connect_timeout=5
```

We will continue testing in order to find a solution to this race condition.

**Comment by rafariossaa** [comment]
Hi,
  I retook this issue, and with version `bitnami/postgresql-repmgr:11.10.0-debian-10-r64` I couldn't reproduce it.
  Could you try with this version ?

**Comment by DanPM** [comment]
> 
> 
> Hi,
> I retook this issue, and with version `bitnami/postgresql-repmgr:11.10.0-debian-10-r64` I couldn't reproduce it.
> Could you try with this version ?

Hello, 
just saw now the issues was closed. 
I will try next week if the problems is actually solved in 11.10.0-debian-10-r64.
Thank you for the good work and support.
